polyglot/download.py

```diff
@@ -0,0 +1,250 @@
+import gzip
+
+from bitsv.network import NetworkAPI
+
+from .bitcom import B, C, BCAT, BCATPART, D, AIP, MAP
+
+class Download(NetworkAPI):
+    def __init__(self, network='main'):
+        super().__init__(network=network)
+
+    # UTILITIES
+    @staticmethod
+    def binary_to_file(binary, file):
+        """Give the pathname to a png, jpg, gif, tiff, bmp
+        writes binary to it
+
+        Example path (for newer users): "C://Users/username/Pictures/my_picture.jpg etc."
+        """
+        with open(file, 'wb') as f:
+            f.write(binary)
+
+    @staticmethod
+    def hex_to_binary(hex):
+        # FIXME - may not just work for any file
+        return bytes.fromhex(hex)
+
+    def scripts_from_txid(self, txid):
+        lst = []
+        tx = self.get_transaction(txid)
+        for output in tx.outputs:
+            lst.append(output.scriptpubkey)
+        return lst
+
+    @staticmethod
+    def binary_to_bsv_string(binary):
+        string = binary.decode('utf-8')
+        if string in ('\0','\t','\n','\x0B','\r',' ',''):
+            string = None
+        return string
+
+    @staticmethod
+    def pushdata_from_script(script):
+        script = Download.hex_to_binary(script)
+        offset = 0
+        data = []
+        while offset < len(script):
+            opcode = script[offset]
+            offset += 1
+            if opcode == 0x00: # OP_0, OP_FALSE
+                data.append(bytes(0))
+            elif opcode <= 0x4b: # short data
+                data.append(script[offset : offset + opcode])
+                offset += opcode
+            elif opcode == 0x4c: # OP_PUSHDATA1
+                length = script[offset]
+                data.append(script[offset + 1 : offset + 1 + length])
+                offset += 1 + length
+            elif opcode == 0x4d: # OP_PUSHDATA2
+                length = script[offset] + script[offset + 1] * 0x100
+                data.append(script[offset + 2 : offset + 2 + length])
+                offset += 2 + length
+            elif opcode == 0x4e: # OP_PUSHDATA4
+                length = script[offset] + script[offset + 1] * 0x100 + script[offset + 2] * 0x10000 + script[offset + 3] * 0x1000000
+                data.append(script[offset + 4 : offset + 4 + length])
+                offset += 4 + length
+            elif opcode == 0x4f: # OP_1NEGATE
+                data.append(bytes([0xff])) # -1 is 0xff in twos complement
+            elif opcode == 0x51: # OP_1, OP_TRUE
+                data.append(bytes([0x01]))
+            elif opcode > 0x50 and opcode <= 0x60: # OP_#
```

Comment:
This looks good.

I think this should technically be:
` elif opcode > 0x51 and opcode <= 0x60: # OP_#`

But the result is the same anyway. I'll change it after merge to get a move on with this...

Just out of interest. For the OP_1NEGATE (0x4f) ->` 0xff` and 0x51 -> `0x01` and OP_# -> `opcode - 0x50`

Do you know if there are any other libraries that take this approach of 'mutating the data' like this.

It makes sense but just would give me more confidence to know if others are doing this aswell. 😄

Reply:
On the same line of thinking, it makes me wonder if `pushdata_from_script` should just not mutate any data at all and directly translate these opcodes as well by appending them to the list of data too... (which may make the returned data useable for a wider range of future use-cases... like parsing more complex scripts in future? (Currently this looks to be taylor made for OP_FALSE OP_RETURN bcat parsing but might shut the door to alternative use cases unless we avoid mutating the original script.

What do you think?
```
            elif opcode == 0x61: # OP_NOP
                pass
            elif opcode == 0x6a: # OP_RETURN
                pass
            elif opcode == 0x80: # OP_RESERVED
                pass
            elif opcode == 0xb0 or (opcode >= 0xb3 and opcode <= 0xb9): # OP_NOP#
                pass
```

Reply:
I'm not asking you to change code - I can merge this but interested to get your feedback to get to an optimal implementation. 😃

Reply:
Transparently, this is the first time I've ever looked at bitcoin opcodes.  I think I copied all the data-related ones from the official wiki.  I just wanted to bang it into functionality.  Really I imagine this code would go in bitsv once we all have enough time and energy to put it there.  Any preferences you have are likely better informed than mine.  Thanks so much for finding time to look at this.